### ISSUE TITLE
[JENKINS-64048] Pods slow to schedule and jobs queueing when large number of parallel jobs with kubernetes 1.27.3

### ISSUE THREAD

**Issue by jenkins-infra-bot** [jenkinsci/kubernetes-plugin#2405]
<p>After update of kubernetes-plugin to 1.27.3 on our production jenkins instance, we started to observe degraded performance of jobs.</p>

<p>Jobs seem to take long time to schedule pods and the queue size is skyrocketing. We are running 100-200 jobs in parallel at times, during our scheduled end to end tests. Normally they get cleared within couple of minutes.</p>

<p>The problem is not apparent until there is spike in jobs scheduled at the same time.</p>

<p>After updated version of kubernetes jobs get stuck and overload the jenkins queue. Our timeout inside of the pipeline kills them after 30 minutes.</p>

<p>We use podTemplate in pipelines to schedule pod, we are not using the "default" jenkins wide templates for pods.</p>

<p>Kubernetes cloud configuration:</p>

<p><span class="image-wrap" style=""><img src="https://issues.jenkins.io/secure/attachment/53071/53071_Screenshot+2020-10-27+at+17.22.40.png" style="border: 0px solid black" /></span></p>

<p> </p>

<p>Some statistics gathered during with the problematic plugin version (1.27.3 - 1.27.4 is mistake in screenshots):</p>

<p><span class="image-wrap" style=""><img src="https://issues.jenkins.io/secure/attachment/53070/53070_Screenshot+2020-10-27+at+17.10.57.png" style="border: 0px solid black" /></span></p>

<p><span class="image-wrap" style=""><img src="https://issues.jenkins.io/secure/attachment/53069/53069_Screenshot+2020-10-27+at+17.12.27.png" style="border: 0px solid black" /></span></p>

<p><span class="image-wrap" style=""><img src="https://issues.jenkins.io/secure/attachment/53068/53068_Screenshot+2020-10-27+at+17.14.55.png" style="border: 0px solid black" /></span></p>

---
<details><summary><i>Originally reported by <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=ldolezal">ldolezal</a>, imported from: <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-64048" target="_blank">Pods slow to schedule and jobs queueing when large number of parallel jobs with kubernetes 1.27.3</a></i></summary>
<i><ul>
<li><b>status</b>: Closed
<li><b>priority</b>: Critical
<li><b>component(s)</b>: kubernetes-plugin
<li><b>resolution</b>: Fixed
<li><b>resolved</b>: 2020-11-16T15:15:51+00:00
<li><b>votes</b>: 0
<li><b>watchers</b>: 4
<li><b>imported</b>: 2025-12-02
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<p>After update of kubernetes-plugin to 1.27.3 on our production jenkins instance, we started to observe degraded performance of jobs.</p>

<p>Jobs seem to take long time to schedule pods and the queue size is skyrocketing. We are running 100-200 jobs in parallel at times, during our scheduled end to end tests. Normally they get cleared within couple of minutes.</p>

<p>The problem is not apparent until there is spike in jobs scheduled at the same time.</p>

<p>After updated version of kubernetes jobs get stuck and overload the jenkins queue. Our timeout inside of the pipeline kills them after 30 minutes.</p>

<p>We use podTemplate in pipelines to schedule pod, we are not using the "default" jenkins wide templates for pods.</p>

<p>Kubernetes cloud configuration:</p>

<p><span class="image-wrap" style=""><img src="https://issues.jenkins.io/secure/attachment/53071/53071_Screenshot+2020-10-27+at+17.22.40.png" style="border: 0px solid black" /></span></p>

<p> </p>

<p>Some statistics gathered during with the problematic plugin version (1.27.3 - 1.27.4 is mistake in screenshots):</p>

<p><span class="image-wrap" style=""><img src="https://issues.jenkins.io/secure/attachment/53070/53070_Screenshot+2020-10-27+at+17.10.57.png" style="border: 0px solid black" /></span></p>

<p><span class="image-wrap" style=""><img src="https://issues.jenkins.io/secure/attachment/53069/53069_Screenshot+2020-10-27+at+17.12.27.png" style="border: 0px solid black" /></span></p>

<p><span class="image-wrap" style=""><img src="https://issues.jenkins.io/secure/attachment/53068/53068_Screenshot+2020-10-27+at+17.14.55.png" style="border: 0px solid black" /></span></p></pre>
</details>
</details>
<details><summary><i>environment</i></summary>

```
Jenkins 2.235.5<br/>
Kubernetes-plugin: 1.27.3
```
</details>
<details><summary><i>4 attachments</i></summary>

- [Screenshot 2020-10-27 at 17.10.57.png](https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-misc/refs/heads/main/attachments/53070/Screenshot%202020-10-27%20at%2017.10.57.png)
  > ![Screenshot 2020-10-27 at 17.10.57.png](https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-misc/refs/heads/main/attachments/53070/Screenshot%202020-10-27%20at%2017.10.57.png)
- [Screenshot 2020-10-27 at 17.12.27.png](https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-misc/refs/heads/main/attachments/53069/Screenshot%202020-10-27%20at%2017.12.27.png)
  > ![Screenshot 2020-10-27 at 17.12.27.png](https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-misc/refs/heads/main/attachments/53069/Screenshot%202020-10-27%20at%2017.12.27.png)
- [Screenshot 2020-10-27 at 17.14.55.png](https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-last/refs/heads/main/attachments/53068/Screenshot%202020-10-27%20at%2017.14.55.png)
  > ![Screenshot 2020-10-27 at 17.14.55.png](https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-last/refs/heads/main/attachments/53068/Screenshot%202020-10-27%20at%2017.14.55.png)
- [Screenshot 2020-10-27 at 17.22.40.png](https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-misc/refs/heads/main/attachments/53071/Screenshot%202020-10-27%20at%2017.22.40.png)
  > ![Screenshot 2020-10-27 at 17.22.40.png](https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-misc/refs/heads/main/attachments/53071/Screenshot%202020-10-27%20at%2017.22.40.png)
</details>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/allan_burdajewicz.png" title="allan_burdajewicz's avatar" /> allan_burdajewicz:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-64048?focusedId=400586&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-400586">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=ldolezal" class="user-hover" rel="ldolezal">ldolezal</a> What version did you upgrade from ?</p>

<p>A significant change in 1.26.3 is that the "Max connections to Kubernetes API" is now respected. Before version 1.26.3, <b>the max number of connection was at most 64</b>, whatever the values provided. </p>

<p>Maybe a high number of concurrent request causes delays. Does the problem still happens if you set the value to 64 ?</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=ldolezal" class="user-hover" rel="ldolezal">ldolezal</a> What version did you upgrade from ?</p>

<p>A significant change in 1.26.3 is that the "Max connections to Kubernetes API" is now respected. Before version 1.26.3, <b>the max number of connection was at most 64</b>, whatever the values provided. </p>

<p>Maybe a high number of concurrent request causes delays. Does the problem still happens if you set the value to 64 ?</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>clabu609:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-64048?focusedId=400770&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-400770">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>We faced an incident like this today on one of our Jenkins instances. We have seven Jenkins instances that schedule about 5000 builds per week each, on a shared large Kubernetes cluster. They were all running k8s plugin version 1.25.4. On Wednesday we upgraded one of them to k8s plugin version 1.27.3. Two days later its queue built up and after hovering at 150 builds in the queue for 2 hours, we were forced to restart it to recover. Prior to this all our Jenkins instances had been running fine without a service restart for 6 months.</p>

<p>I think this might be the pull request <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=allan_burdajewicz" class="user-hover" rel="allan_burdajewicz">allan_burdajewicz</a> is referring to with the significant change: <a href="https://github.com/jenkinsci/kubernetes-plugin/pull/824" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/kubernetes-plugin/pull/824</a>. It looks like it is being rework by this pull request: <a href="https://github.com/jenkinsci/kubernetes-plugin/pull/835" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/kubernetes-plugin/pull/835</a>. </p>

<p>Ping <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=seakip18" class="user-hover" rel="seakip18">seakip18</a> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=wolniewicz" class="user-hover" rel="wolniewicz">wolniewicz</a>.</p>
</pre>
</details>

</ul>
</details>

<p>We faced an incident like this today on one of our Jenkins instances. We have seven Jenkins instances that schedule about 5000 builds per week each, on a shared large Kubernetes cluster. They were all running k8s plugin version 1.25.4. On Wednesday we upgraded one of them to k8s plugin version 1.27.3. Two days later its queue built up and after hovering at 150 builds in the queue for 2 hours, we were forced to restart it to recover. Prior to this all our Jenkins instances had been running fine without a service restart for 6 months.</p>

<p>I think this might be the pull request <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=allan_burdajewicz" class="user-hover" rel="allan_burdajewicz">allan_burdajewicz</a> is referring to with the significant change: <a href="https://github.com/jenkinsci/kubernetes-plugin/pull/824" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/kubernetes-plugin/pull/824</a>. It looks like it is being rework by this pull request: <a href="https://github.com/jenkinsci/kubernetes-plugin/pull/835" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/kubernetes-plugin/pull/835</a>. </p>

<p>Ping <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=seakip18" class="user-hover" rel="seakip18">seakip18</a> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=wolniewicz" class="user-hover" rel="wolniewicz">wolniewicz</a>.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>clabu609:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-64048?focusedId=400812&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-400812">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I think this could be a duplicate of <a href="https://issues.jenkins.io/browse/JENKINS-63705" title="Concurrency limit not calculated over template anymore in 1.27.1" class="issue-link" data-issue-key="JENKINS-63705"><del>JENKINS-63705</del></a></p>
</pre>
</details>

</ul>
</details>

<p>I think this could be a duplicate of <a href="https://issues.jenkins.io/browse/JENKINS-63705" title="Concurrency limit not calculated over template anymore in 1.27.1" class="issue-link" data-issue-key="JENKINS-63705"><del>JENKINS-63705</del></a></p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/vlatombe.png" title="vlatombe's avatar" /> vlatombe:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-64048?focusedId=401219&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-401219">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Should be fixed with <a href="https://github.com/jenkinsci/kubernetes-plugin/pull/880" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/kubernetes-plugin/pull/880</a> =&gt; 1.27.5 and above.</p>
</pre>
</details>

</ul>
</details>

<p>Should be fixed with <a href="https://github.com/jenkinsci/kubernetes-plugin/pull/880" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/kubernetes-plugin/pull/880</a> => 1.27.5 and above.</p>
